hal/src/thumbv7em/pukcc.rs

```diff
@@ -0,0 +1,579 @@
+            // it is placed at the end, idea is that algorithm will use whatever amount of
+            // memory it needs
+            (workspace, &[0_u8; 0])
+        };
+        let mut pukcl_params = c_abi::PukclParams::default();
+        unsafe {
+            let mut service_params = &mut pukcl_params.params.ZpEcDsaGenerateFast;
+            service_params.nu1ModBase = modulo_p.pukcc_base();
+            service_params.nu1CnsBase = cns.pukcc_base();
+            service_params.u2ModLength = C::MOD_LENGTH as u16;
+            service_params.nu1ScalarNumber = k_cr.pukcc_base();
+            service_params.nu1OrderPointBase = order_point.pukcc_base();
+            service_params.nu1PrivateKey = private_key_cr.pukcc_base();
+            service_params.nu1HashBase = hash_cr.pukcc_base();
+            service_params.u2ScalarLength = C::SCALAR_LENGTH;
+            service_params.nu1PointABase = base_point_a_x.pukcc_base();
+            service_params.nu1ABase = a_curve.pukcc_base();
+            service_params.nu1Workspace = workspace.pukcc_base();
+        }
+
+        unsafe { c_abi::ZpEcDsaGenerateFast::call(&mut pukcl_params) };
+
+        match pukcl_params.header.u2Status.into() {
+            PukclReturnCode::Ok => {}
+            error_code => return Err(EcdsaSignFailure::ServiceFailure(error_code)),
+        };
+
+        // Generated signature R part is written to base point X coordinate memory.
+        // Generated signature S part is written to base point Y coordinate memory.
+        // Base point Z coordinate should be zero.
+
+        if !base_point_a_z.iter().all(|&el| el == 0) {
+            return Err(EcdsaSignFailure::BasePointZCoordinateIsNotZero);
+        }
+
+        // Copying signature back from the CryptoRAM while ignoring irrelevant padding.
+        signature
+            .iter_mut()
+            .zip(
+                base_point_a_x
+                    .iter()
+                    .rev()
+                    .skip(4)
+                    .chain(base_point_a_y.iter().rev().skip(4)),
+            )
+            .for_each(|(target_iter, source_iter)| *target_iter = *source_iter);
+
+        Ok(())
+    }
+
+    /// Service verifying an ECDSA signature.
+    ///
+    /// GF(p) service. GF(2^n) variant is not implemented -- use low-level API.
+    ///
+    /// Input parameters:
+    /// - `signature`: `&[u8]` of length `2 * `[`Curve::SCALAR_LENGTH`]
+    ///     - Signature that is being verified
+    /// - `hash`: `&[u8]` of length [`Curve::SCALAR_LENGTH`]
+    ///     - Hash of a message that is signed.
+    /// - `public_key`: `&[u8]` of length [`Curve::SCALAR_LENGTH`]
+    ///     - Public key used for a signature verification.
+    ///
+    /// Return value:
+    /// - `Result::Ok`
+    ///     - Signature is valid against chosen `hash` and `public_key`
+    /// - `Result::Err`
+    ///     - Possible failure scenarios are encapsulated in a
```

Comment:
Is it possible that this could be a padding/error oracle? In other words, could different kinds of failures indicate information about the key to an attacker?

If so, it might be worth adding a disclaimer to not expose the details from `EcdsaSignatureVerificationFailure` outside of debugging.

Reply:
No. According to 43.3.6.12.7 in datasheet it is either "ok" or "wrong signature". No specifics. For signature generation (43.3.6.11.7) it is either "ok" or "wrong select number" which is `k`.